cmake: STACK_CANARIES: Re-write how compatibility is checked

SebastianBoe opened the pull request:
This re-write fixes several issues with how we were checking for
compatibility with the canary flags:

If a compiler did not support any stack canaries, but a user had still
enabled them, this would silently result in no canaries being added
instead of error-ing out as it should have.

Compatiblity for '-mstack-protector-guard=global' was unnecessarily
being tested twice, once in 'check_c_compiler_flag' and once in
'zephyr_cc_option'.

Compatibility was being tested with 'check_c_compiler_flag' which is
slower than 'zephyr_check_compiler_flag' because it does not cache
it's results.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: STACK_CANARIES: Re-write how compatibility is checked

This re-write fixes several issues with how we were checking for
compatibility with the canary flags:

If a compiler did not support any stack canaries, but a user had still
enabled them, this would silently result in no canaries being added
instead of error-ing out as it should have.

Compatiblity for '-mstack-protector-guard=global' was unnecessarily
being tested twice, once in 'check_c_compiler_flag' and once in
'zephyr_cc_option'.

Compatibility was being tested with 'check_c_compiler_flag' which is
slower than 'zephyr_check_compiler_flag' because it does not cache
it's results.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

tyrantlizard77 reviewed (dismissed):
> instead of error-ing out as it should have.

You misunderstand what my patch was for. The use of check_c_compiler_flag() was intentional.
Earlier versions of GCC always used the global guard.
Sometime in GCC 7.x, they defaulted to thread-local storage, which we don't support. You have to pass -mstack-protector-guard=global or it doesn't work. But if you pass that in GCC 6.x, it breaks the build. So that's why there is a test before we try to set the second option.

SebastianBoe commented:
> Sometime in GCC 7.x, they defaulted to thread-local storage, which we don't support. You have to pass -mstack-protector-guard=global or it doesn't work. But if you pass that in GCC 6.x, it breaks the build. So that's why there is a test before we try to set the second option.

As this patch is written. '-mstack-protector-guard=global' is passed when the toolchain supports it, and ''-fstack-protector-all' is passed unconditionally.

This gives us the desired behaviour on both old and new versions of GCC.

As it was written before, '-fstack-protector-all' was only applied when it was supported by the toolchain, but this is not desired behaviour.

SebastianBoe commented:
recheck

SebastianBoe commented:
@andrewboie : Please re-assess the review when convenient. Be aware that *_cc_* will check for compiler compatiblity.

SebastianBoe commented:
@andrewboie : Your NACK is blocking the PR. Your concerns have been addressed.

Please re-asses the PR.

SebastianBoe commented:
@nashif : The reviewer is unresponsive.

I would like to see this reviewed and merged in reasonable time.

SebastianBoe committed:
Kconfig: STACK_CANARIES: Correct the help text

The help text has been stating that CONFIG_STACK_CANARIES will
silently be ignored when the compiler does not support them. But this
is not the desired behaviour of CONFIG_STACK_CANARIES[1].

This patch corrects the help text to state that an error will occur if
this feature is enabled, but not supported.

[1] "I would much rather see the build break if someone tries to
enable the stack canaries, and the compiler doesn't support
it. Because what happens now is that if someone enables this option,
and there is no support, the build will succeed but there are no
actual stack canaries in place, and unless the user is paying close
attention to the cmake test output they will have no idea."
--
https://github.com/zephyrproject-rtos/zephyr/issues/5019

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
@carlescufi : This is in need of review.

nashif commented:
hmm, something weird, now with current zephyr SDK we get

`-fstack-protector-all -mstack-protector-guard=global` when building, I thought this was not supported with gcc 6.x.

carlescufi reviewed (approved):
I have no objections to this

SebastianBoe commented:
> -fstack-protector-all -mstack-protector-guard=global when building, I thought this was not supported with gcc 6.x.

As did I. But evidently it is possible to build a 6.2.0 toolchain with -mstack-protector-guard=global support because we have two 6.2.0 toolchains in SDK 0.9.5, an ARM toolchain and an X86 toolchain. And the ARM toolchain does not support 'guard=global', but the x86 toolchain does.

In any case, as the build scripts are now written, we always get the desired behaviour.
